src/libs/mps_comm/machine.cpp

```diff
@@ -0,0 +1,479 @@
+                 unsigned short int machine_type,
+                 std::string        ip,
+                 unsigned short     port,
+                 ConnectionMode     connection_mode)
+: abort_operation_(false),
+  name_(name),
+  machine_type_(machine_type),
+  ip_(ip),
+  port_(port),
+  connection_mode_(connection_mode),
+  shutdown_(false),
+  heartbeat_active_(false)
+{
+	initLogger();
+	worker_thread_ = std::thread(&Machine::dispatch_command_queue, this);
+}
+
+void Machine::dispatch_command_queue()
+{
+	std::unique_lock<std::mutex> lock(command_queue_mutex_);
+	// Always process the queue before shutdown
+	while (!shutdown_ || !command_queue_.empty()) {
+		queue_condition_.wait(lock, [this] { return !command_queue_.empty() || shutdown_; });
+    if (!command_queue_.empty()) {
+			auto command = command_queue_.front();
+			command_queue_.pop();
+			lock.unlock();
+			command();
+			using namespace std::chrono_literals;
+			std::this_thread::sleep_for(40ms);
+			lock.lock();
+    }
+	}
+}
+
+void Machine::heartbeat()
+{
+  heartbeat_active_ = true;
+  using namespace std::chrono_literals;
+  while (!shutdown_) {
+    send_command(COMMAND_NOTHING, 0, 0, 1);
+    std::this_thread::sleep_for(1s);
+  }
+  heartbeat_active_ = false;
+}
+
+void Machine::mock_callback(OpcUtils::MPSRegister reg, OpcUtils::ReturnValue *ret)
+{
+	for (auto &cb : callbacks_) {
+		if (std::get<1>(cb) == reg) {
+			SubscriptionClient::ReturnValueCallback fct = std::get<0>(cb);
+			fct(ret);
+		}
+	}
+}
+
+void Machine::mock_callback(OpcUtils::MPSRegister reg, bool ret)
+{
+  OpcUtils::ReturnValue ret_val;
+  ret_val.bool_s = ret;
+  return mock_callback(reg, &ret_val);
+}
+
+void Machine::send_command(unsigned short command, unsigned short payload1, unsigned short payload2, int timeout, unsigned char status, unsigned char error) {
+	std::function<void(void)> call = [this, command, payload1, payload2, status, error] {
+		std::unique_lock<std::mutex> lock(command_mutex_);
+		bool                         success;
+		do {
+			try {
+				logger->info("Sending command: {} {} {} {}", command, payload1, payload2, status);
+        if (connection_mode_ == MOCKUP) {
+          if (command > 100) {
+            mock_callback(OpcUtils::MPSRegister::STATUS_BUSY_IN, true);
+            std::this_thread::sleep_for(mock_busy_duration_);
+            mock_callback(OpcUtils::MPSRegister::STATUS_BUSY_IN, false);
+            mock_callback(OpcUtils::MPSRegister::STATUS_READY_IN, true);
+            std::this_thread::sleep_for(mock_ready_duration_);
+            mock_callback(OpcUtils::MPSRegister::STATUS_READY_IN, false);
+          }
+          return;
+        }
+				OpcUtils::MPSRegister registerOffset;
+				if (command < Station::STATION_BASE)
+					registerOffset = OpcUtils::MPSRegister::ACTION_ID_BASIC;
+				else
+					registerOffset = OpcUtils::MPSRegister::ACTION_ID_IN;
+
+				bool                  statusBit = (bool)(status & Status::STATUS_BUISY);
```

Comment:
```suggestion
				bool                  statusBit = (bool)(status & Status::STATUS_BUSY);
```
That looks like a typo.

Reply:
Thanks for pointing those out, I fixed all of them.